http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/RunnableRequest.java

```diff
@@ -111,13 +112,43 @@ private URI endpoint() {
     private void handleFailure(Supplier<String> msg, Throwable cause) {
         log.error(msg.get(), cause);
         runAndLogError("Exception thrown from AsyncResponseHandler",
-            () -> context.handler().exceptionOccurred(cause));
+            () -> context.handler().exceptionOccurred(modifyHighBurstTrafficException(cause)));
         if (channel != null) {
-            runAndLogError("Unable to release channel back to the pool.",
-                () -> closeAndRelease(channel));
+            runAndLogError("Unable to release channel back to the pool.", () -> closeAndRelease(channel));
         }
     }
 
+    private Throwable modifyHighBurstTrafficException(Throwable originalCause) {
+        String originalMessage = originalCause.getMessage();
+        String newMessage = null;
+
+        if (originalCause instanceof TimeoutException &&
+            originalMessage.contains("Acquire operation took longer")) {
+
+            newMessage = "Timed out waiting for a connection to be released to the pool. "
+                         + "Consider increasing your max connections or pool lease timeout.";
+
+        } else if (originalCause instanceof IllegalStateException &&
+                   originalMessage.contains("Too many outstanding acquire operations")) {
+
+            newMessage = new StringBuilder()
+                             .append("Request rate is too high. You can consider taking any of the following actions to mitigate"
+                                     + " the issue: Increase max connections, increase max pending acquire count, "
+                                     + "decrease pool lease timeout, or slowing down the requests rate.\n")
+                             .append("SDK default max connections is generally less for high burst traffic, you can increase "
+                                     + "max connections. If you already have high max connections value, increasing the max "
+                                     + "pending acquire count allows extra requests to be buffered. If this is not possible, "
+                                     + "consider decreasing the lease timeout which will fail fast the application.\n")
+                             .append("If none of these options work, consider slowing the rate of requests sent (for example "
+                                     + "using a semaphore)")
```

Comment:
Do you think we should include similar details for the acquire operation taking too long?

For this one, I'd maybe say:

Maximum pending connection acquisitions exceeded. The request rate is too high for the client to keep up. 

Consider taking any of the following actions to mitigate the issue: increase max connections, increase max pending acquire count, decrease pool lease timeout, or slowing the request rate.

Increasing the max connections can increase client throughput (unless the network interface is already fully utilized), but can eventually start to hit operation system limitations on the number of file descriptors used by the process. If you already are fully utilizing your network interface or cannot further increase your connection count, increasing the pending acquire count allows extra requests to be buffered by the client, but can cause additional request latency and higher memory usage. If your request latency or memory usage is already too high, decreasing the lease timeout will allow requests to fail more quickly, reducing the number of pending connection acquisitions, but likely won't decrease the total number of failed requests.

If the above mechanisms are not able to fix the issue, try "smoothing out" your requests so that large traffic "bursts" cannot overload the client, being more efficient with the number of times you need to call AWS, or by increasing the number of hosts sending requests.

Reply:
Updated